optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraint.java

```diff
@@ -57,13 +60,52 @@ public DroolsConstraint(DroolsConstraintFactory<Solution_> constraintFactory,
         return positive ? constraintWeight : constraintWeight.negate();
     }
 
-    public Rule createRule(Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
-        List<RuleItemBuilder<?>> ruleItemBuilderList = new ArrayList<>(fromStreamList.size());
-        for (DroolsFromUniConstraintStream<Solution_, Object> fromStream : fromStreamList) {
-            fromStream.createRuleItemBuilders(ruleItemBuilderList, scoreHolderGlobal);
+    private Set<DroolsAbstractConstraintStream<Solution_>> assembleAllStreams() {
+        final Set<DroolsAbstractConstraintStream<Solution_>> streams = new LinkedHashSet<>();
+        for (DroolsAbstractConstraintStream<Solution_> fromStream: fromStreamList) {
+            streams.addAll(assembleAllStreams(fromStream));
         }
-        return PatternDSL.rule(constraintPackage, constraintName)
-                .build(ruleItemBuilderList.toArray(new RuleItemBuilder<?>[0]));
+        return streams;
+    }
+
+    private Collection<DroolsAbstractConstraintStream<Solution_>> assembleAllStreams(
+            DroolsAbstractConstraintStream<Solution_> parent) {
+        final Set<DroolsAbstractConstraintStream<Solution_>> streams = new LinkedHashSet<>();
+        streams.add(parent);
+        for (DroolsAbstractConstraintStream<Solution_> child: parent.getChildStreams()) {
+            streams.addAll(assembleAllStreams(child));
+        }
+        return streams;
+    }
+
+    /**
+     * Retrieves Drools rules required to process this constraint.
+     *
+     * @param ruleLibrary never null. Cache of rules already generated by previous constraints. This method will use
+     * {@link Map#computeIfAbsent(Object, Function)} to add rules from {@link DroolsAbstractConstraintStream}s which we
+     * have not yet processed. This way, we will only generate one rule per stream, allowing Drools to reuse the
+     * computations.
+     * @param scoreHolderGlobal never null. The Drools global used to track changes to score within rule consequences.
+     * @return never null. List of all Drools {@link Rule}s required by this constraint. May update ruleLibrary as a
+     * side effect.
+     */
+    public List<Rule> createRules(Map<DroolsAbstractConstraintStream<Solution_>, Rule> ruleLibrary,
+            Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
```

Comment:
I presume the ruleLibrary 's lifecycle scope is inclosed in the ScoreDirectorFactory creation?
AKA, building 2 ScoreDirectorFactory's means they both have their own ruleLibrary? (That's a good thing, there is only one ScoreDirectorFactory build anyway normally).

Reply:
yes, just confirmed it